src/hugr/rewrite/inline_dfg.rs

```diff
@@ -0,0 +1,281 @@
+//! A rewrite that inlines a DFG node, moving all children
+//! of the DFG except Input+Output into the DFG's parent,
+//! and deleting the DFG along with its Input + Output
+
+use super::Rewrite;
+use crate::ops::handle::{DfgID, NodeHandle};
+use crate::ops::OpType;
+use crate::{Direction, IncomingPort, Node, PortIndex};
+
+/// Structure identifying an `InlineDFG` rewrite from the spec
+pub struct InlineDFG(pub DfgID);
+
+/// Errors from an [InlineDFG] rewrite.
+#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
+pub enum InlineDFGError {
+    /// Node to inline was not a DFG. (E.g. node has been overwritten since the DfgID originated.)
+    #[error("Node {0} was not a DFG")]
+    NotDFG(Node),
+    /// DFG has no parent (is the root)
+    #[error("Node did not have a parent into which to inline")]
+    NoParent,
+    /// DFG has other edges (i.e. Order edges) incoming/outgoing.
+    /// (We don't support such as the new endpoints for such edges is not clear.)
+    #[error("DFG node had non-dataflow edges in direction {0:?}")]
+    HasOtherEdges(Direction),
+}
+
+impl Rewrite for InlineDFG {
+    /// Returns the removed nodes: the DFG, and its Input and Output children,
+    type ApplyResult = [Node; 3];
+    type Error = InlineDFGError;
+
+    type InvalidationSet<'a> = <[Node; 1] as IntoIterator>::IntoIter;
+
+    const UNCHANGED_ON_FAILURE: bool = true;
+
+    fn verify(&self, h: &impl crate::HugrView) -> Result<(), Self::Error> {
+        let n = self.0.node();
+        let op @ OpType::DFG { .. } = h.get_optype(n) else {
+            return Err(InlineDFGError::NotDFG(n));
+        };
+        if h.get_parent(n).is_none() {
+            return Err(InlineDFGError::NoParent);
+        };
+
+        for d in Direction::BOTH {
+            if op
+                .other_port(d)
+                .is_some_and(|p| h.linked_ports(n, p).next().is_some())
+            {
+                return Err(InlineDFGError::HasOtherEdges(d));
+            };
+        }
+
+        Ok(())
+    }
+
+    fn apply(self, h: &mut impl crate::hugr::HugrMut) -> Result<Self::ApplyResult, Self::Error> {
```

Comment:
Do we expect this to work when the DFG has non-local edges either incoming or outgoing? I can't see why it wouldn't, in either case -- but perhaps worth adding a test.

(Incoming non-local edges from the surrounding DFG would just become local, but I don't think that needs any special handling?)